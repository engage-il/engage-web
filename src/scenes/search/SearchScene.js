// @flow
import React, { Component } from 'react'
import { createRefetchContainer, graphql } from 'react-relay'
import type { RelayRefetchProp } from 'react-relay'
import { withRouter } from 'react-router-dom'
import type { ContextRouter } from 'react-router-dom'
import { throttle } from 'lodash'
import { constants } from './searchRoute'
import { Intro, SearchField, BillsList, LoadingIndicator } from './components'
import { stylesheet, colors, mixins } from 'shared/styles'
import type { Viewer } from 'shared/types'

let SearchScene = class SearchScene extends Component {
  props: {
    viewer: ?Viewer,
    relay: RelayRefetchProp
  } & ContextRouter

  state = {
    query: constants.query,
    disableAnimations: false
  }

  // events
  searchFieldDidChange = (query: string) => {
    this.setState({ query })
    this.filterBillsForQuery(query)
  }

  // helpers
  filterBillsForQuery: (query: string) => void = throttle((query: string) => {
    this.setState({ disableAnimations: true })
    this.props.relay.refetch({ query }, null, (error: ?Error) => {
      if (error) {
        console.error(`error updaing query: ${error.toString()}`)
      }

      // completion comes back before render
      requestAnimationFrame(() => {
        this.setState({ disableAnimations: false })
      })
    })
  }, 300)

  // lifecycle
  render () {
    const { viewer } = this.props
    const { query, disableAnimations } = this.state

    return <div {...rules.container}>
      <div {...rules.header}>
        <div {...rules.background} />
        <Intro styles={rules.section} />
        <SearchField
          styles={rules.section}
          value={query}
          onChange={this.searchFieldDidChange}
        />
      </div>
      <div {...rules.content}>
        <div {...rules.indicator}>
          <LoadingIndicator isLoading={!viewer} />
        </div>
        {viewer && <BillsList
          viewer={viewer}
          animated={!disableAnimations}
        />}
      </div>
    </div>
  }
}

SearchScene = createRefetchContainer(withRouter(SearchScene),
  graphql`
    fragment SearchScene_viewer on Viewer {
      bills(
        first: $count, after: $cursor,
        query: $query, from: $startDate, to: $endDate
      ) {
        edges { node { id } }
      }
      ...BillsList_viewer
    }
  `,
  graphql`
    query SearchSceneQuery(
      $count: Int!, $cursor: String!,
      $query: String!, $startDate: Time!, $endDate: Time!
    ) {
      viewer {
        ...SearchScene_viewer
      }
    }
  `
)

const rules = stylesheet({
  container: {
    ...mixins.column
  },
  header: {
    position: 'relative',
    marginBottom: 30,
    ...mixins.mobile({
      marginBottom: 15
    })
  },
  background: {
    position: 'absolute',
    top: -30,
    bottom: 0,
    left: -30,
    width: '100vw',
    zIndex: -1,
    backgroundColor: colors.backgroundAccent,
    ...mixins.mobile({
      left: -15
    })
  },
  section: {
    marginBottom: 30,
    ...mixins.mobile({
      marginBottom: 15
    })
  },
  content: {
    ...mixins.column,
    position: 'relative'
  },
  loadMoreButton: {
    alignSelf: 'center',
    marginTop: 30,
    ...mixins.mobile({
      marginTop: 20
    })
  },
  indicator: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    ...mixins.column
  }
})

export { SearchScene }
