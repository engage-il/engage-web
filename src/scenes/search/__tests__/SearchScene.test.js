/* eslint-env jest */
import React from 'react'
import { shallow, mount } from 'enzyme'
import { SearchScene } from '../SearchScene'
import { relayRefetchProp } from 'mocks/relayProps'

const { anything } = expect

// subject
let subject
let viewer

function loadSubject (options = { mount: false }) {
  const renderer = options.mount ? mount : (component) => shallow(component).dive().dive()
  subject = renderer(<SearchScene viewer={viewer} />)
}

const element = {
  list: () => subject.find('BillsList'),
  searchField: () => subject.find('SearchField'),
  indicator: () => subject.find('LoadingIndicator')
}

// specs
beforeEach(() => {
  // TODO: build rosie.js factories
  viewer = {
    bills: {
      edges: []
    }
  }
})

afterEach(() => {
  subject = null
})

describe('#state', () => {
  beforeEach(loadSubject)

  it('initially has a blank query', () => {
    expect(subject).toHaveState('query', '')
  })
})

describe('#render', () => {
  it('shows the search field with the current query', () => {
    loadSubject()
    subject.setState({ query: 'foo' })
    expect(element.searchField()).toHaveValue('foo')
  })

  it('shows the bills list', () => {
    loadSubject()
    const list = element.list()
    expect(list).toBePresent()
    expect(list).toHaveProp('viewer', viewer)
  })

  it('hides the loading indicator', () => {
    loadSubject()
    expect(element.indicator()).toHaveProp('isLoading', false)
  })

  describe('when loading', () => {
    it('shows the loading indicator', () => {
      viewer = null
      loadSubject()
      expect(element.indicator()).toHaveProp('isLoading', true)
    })

    it('hides the bills list', () => {
      viewer = null
      loadSubject()
      expect(element.list()).toBeEmpty()
    })
  })
})

describe('on search field change', () => {
  it('updates the visible query', () => {
    loadSubject()
    element.searchField().simulate('change', 'foo')
    expect(subject).toHaveState('query', 'foo')
  })

  it('disables animations', () => {
    loadSubject()
    element.searchField().simulate('change', 'foo')
    expect(element.list()).toHaveProp('animated', false)
  })

  it('refetches the relay query', () => {
    loadSubject()
    element.searchField().simulate('change', 'foo')
    expect(relayRefetchProp.refetch).toHaveBeenLastCalledWith({ query: 'foo' }, null, anything())
  })
})

describe('the relay container', () => {
  it('exists', () => {
    expect(SearchScene.container).toBeTruthy()
  })
})
