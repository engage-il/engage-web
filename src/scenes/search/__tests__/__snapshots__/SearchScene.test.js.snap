// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#render normally 1`] = `
<div
  data-css-1szwbq4=""
>
  <div
    data-css-p2q3yg=""
  >
    <div
      data-css-tyhdvi=""
    />
    <Intro
      styles={
        Object {
          "data-css-unxum8": "",
        }
      }
    />
    <SearchField
      onChange={[Function]}
      styles={
        Object {
          "data-css-unxum8": "",
        }
      }
      value=""
    />
  </div>
  <div
    data-css-7551rz=""
  >
    <div
      data-css-194orap=""
    >
      <LoadingIndicator
        isLoading={false}
      />
    </div>
    <BillsList
      animated={true}
      viewer={
        Object {
          "bills": Object {
            "edges": Array [],
          },
        }
      }
    />
  </div>
</div>
`;

exports[`#render renders properly when loading 1`] = `
<div
  data-css-1szwbq4=""
>
  <div
    data-css-p2q3yg=""
  >
    <div
      data-css-tyhdvi=""
    />
    <Intro
      styles={
        Object {
          "data-css-unxum8": "",
        }
      }
    />
    <SearchField
      onChange={[Function]}
      styles={
        Object {
          "data-css-unxum8": "",
        }
      }
      value=""
    />
  </div>
  <div
    data-css-7551rz=""
  >
    <div
      data-css-194orap=""
    >
      <LoadingIndicator
        isLoading={true}
      />
    </div>
  </div>
</div>
`;

exports[`#render with query 1`] = `
<div
  data-css-1szwbq4=""
>
  <div
    data-css-p2q3yg=""
  >
    <div
      data-css-tyhdvi=""
    />
    <Intro
      styles={
        Object {
          "data-css-unxum8": "",
        }
      }
    />
    <SearchField
      onChange={[Function]}
      styles={
        Object {
          "data-css-unxum8": "",
        }
      }
      value="foo"
    />
  </div>
  <div
    data-css-7551rz=""
  >
    <div
      data-css-194orap=""
    >
      <LoadingIndicator
        isLoading={false}
      />
    </div>
    <BillsList
      animated={true}
      viewer={
        Object {
          "bills": Object {
            "edges": Array [],
          },
        }
      }
    />
  </div>
</div>
`;
